### ISSUE TITLE
请问训练完成后如何使用测试集进行测试呢？

### ISSUE THREAD

**Issue by yuyukuy534** [zZhiG/crackmer#2]
请问训练完成后如何使用测试集进行测试呢？test.py代码可以发一下吗，非常感谢您

**Comment by zZhiG** [comment]
我们上传了一个可用于测试的代码。
该代码会将所采用测试集的原始图像、预测结果、标签保存为.npy格式到相应文件夹下。
我们可以重新加载.npy文件去计算各项指标、并显示结果等。
您也可以根据自己的需求修改测试代码，例如添加实时的可视化，修改测试过程中对图像的处理过程等。

**Comment by yuyukuy534** [comment]
> 我们上传了一个可用于测试的代码。 该代码会将所采用测试集的原始图像、预测结果、标签保存为.npy格式到相应文件夹下。 我们可以重新加载.npy文件去计算各项指标、并显示结果等。 您也可以根据自己的需求修改测试代码，例如添加实时的可视化，修改测试过程中对图像的处理过程等。
收到，请问我直接在测试的时候把结果打印出来得到的结果为啥和论文中的结果相差很多，是哪里写的不对吗？求大佬帮忙看下代码，代码和结果如下（加载的权重是训练80轮的权重），非常感谢您
import torch
import numpy as np
from torch.utils.data import DataLoader
import os
from src.Net import Net
from utils.dataloader import Datases_loader as dataloader

def compute_confusion_matrix(precited,expected):
    part = np.logical_xor(precited, expected)
    pcount = np.bincount(part)
    tp_list = list(np.logical_and(precited, expected))
    fp_list = list(np.logical_and(precited, np.logical_not(expected)))
    tp = tp_list.count(1)
    fp = fp_list.count(1)                  
    tn = pcount[0] - tp
    fn = pcount[1] - fp
    return tp, fp, tn, fn

def compute_indexes(tp, fp, tn, fn):
    accuracy = (tp+tn) / (tp+tn+fp+fn)
    precision = tp / (tp+fp)
    recall = tp / (tp+fn)
    F1 = (2*precision*recall) / (precision+recall)
    mIoU = np.mean(tp/(tp+fp+fn))
    return accuracy, precision, recall, F1, mIoU
    
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
batchsz = 1
model = Net().to(device)
savedir = '/home/gdg/crackmer-main/weights/model79.pth'

imgdir = '/home/gdg/crackmer-main/DeepCrack/test_img'
labdir = '/home/gdg/crackmer-main/DeepCrack/test_lab'
imgsz = 512

dataset = dataloader(imgdir, labdir, imgsz, imgsz)
testsets = DataLoader(dataset, batch_size=batchsz, shuffle=False)

def test():
    model.load_state_dict(torch.load(savedir))
    tp, tn, fp, fn = 0, 0, 0, 0
    for idx, samples in enumerate(testsets):
        img, lab = samples['image'], samples['mask']
        img, lab = img.to(device), lab.to(device)

        pred = model(img)
        
        p = pred.reshape(-1)
        p[p >= 0.] = 1
        p[p < 0.] = 0
        t = lab.reshape(-1)
        tp_, fp_, tn_, fn_ = compute_confusion_matrix(p.detach().cpu().numpy(), t.detach().cpu().numpy())
        tp = tp + tp_
        fp = fp + fp_
        tn = tn + tn_
        fn = fn + fn_

    accuracy_, precision_, recall_, F1_,mIoU_ = compute_indexes(tp, fp, tn, fn)
    
    print('accuracy:',accuracy_)
    print('precision:',precision_)
    print('recall:',recall_)
    print('F1:',F1_)
    print('mIoU:',mIoU_)
    
if __name__ == '__main__':
    test()

结果为：
<img width="252" alt="e3f5a80a21a1358b75de66ab72731d5" src="https://github.com/zZhiG/crackmer/assets/41134979/cfd20120-0db6-4210-a410-b3686aee5308">

**Comment by yuyukuy534** [comment]
> 我们上传了一个可用于测试的代码。 该代码会将所采用测试集的原始图像、预测结果、标签保存为.npy格式到相应文件夹下。 我们可以重新加载.npy文件去计算各项指标、并显示结果等。 您也可以根据自己的需求修改测试代码，例如添加实时的可视化，修改测试过程中对图像的处理过程等。

收到，请问我直接在测试的时候把结果打印出来得到的结果为啥和论文中的结果相差很多，是哪里写的不对吗？求大佬帮忙看下代码，代码和结果如下（加载的权重是训练80轮的权重），非常感谢您
![image](https://github.com/zZhiG/crackmer/assets/41134979/0d839998-8b2c-4669-aa84-2cb3593767ae)
测试结果如下：
<img width="252" alt="e3f5a80a21a1358b75de66ab72731d5" src="https://github.com/zZhiG/crackmer/assets/41134979/297b26c4-970a-493c-9775-386883edd083">

**Comment by zZhiG** [comment]
可能存在以下原因：
我们在测试时没有对图像进行数据增强。
我们计算是裂缝和背景两类的miou。
采用数据集划分方式的不同。

**Comment by yuyukuy534** [comment]
> 可能存在以下原因： 我们在测试时没有对图像进行数据增强。 我们计算是裂缝和背景两类的miou。 采用数据集划分方式的不同。

好的，请问大佬如何计算裂缝和背景两类的mIoU呢？求指导

**Comment by zZhiG** [comment]
是的，模型预测会得到裂缝和背景两类，我们对这两部分都进行了计算。
而有一些论文中则只会计算裂缝这一类。
